codeworld-api/src/CodeWorld/Driver.hs

```diff
@@ -1161,7 +1214,64 @@ run initial stepHandler eventHandler drawHandler = do
     t0 <- getTime
     nullFrame <- makeStableName undefined
     initialStateName <- makeStableName $! initial
-    go t0 nullFrame initialStateName True
+
+    forkIO $ go t0 nullFrame initialStateName True
+    return (sendEvent, getState)
+
+data StateWrapper s = StateWrapper Bool s
+data EventWrapper = NormalEvent Event | PauseEvent Bool
+
+-- Wraps the event and state from run so they can be paused by pressing the Inspect
+-- button.
+runInspect :: s -> (Double -> s -> s) -> (Event -> s -> s) -> (s -> Picture) -> IO ()
+runInspect initial stepHandler eventHandler drawHandler = do
+    Just window <- currentWindow
+    Just doc <- currentDocument
+    Just canvas <- getElementById doc ("screen" :: JSString)
+
+    let initialWrapper = StateWrapper False initial
+        stepHandlerWrapper dt (StateWrapper paused state) = case paused of
+            True  -> StateWrapper True state
+            False -> StateWrapper False (stepHandler dt state)
+        eventHandlerWrapper e (StateWrapper p s) = case e of
+            NormalEvent event -> case p of
+                True  -> StateWrapper True s
+                False -> StateWrapper False (eventHandler event s)
+            PauseEvent paused -> StateWrapper paused s
+        drawHandlerWrapper (StateWrapper _ s) = drawHandler s
+
+    (sendEvent, getState) <-
+        run initialWrapper stepHandlerWrapper eventHandlerWrapper drawHandlerWrapper
+
+    onEvents canvas (sendEvent . NormalEvent)
+    inspectFromIO (sendEvent . PauseEvent) $ drawHandlerWrapper <$> getState
+
+-- Allows pictures to be inspected using a built-in pause button
+runPauseable :: s -> (Double -> s -> s) -> (Event -> s -> s) -> (s -> Picture) -> (s -> Bool) -> IO ()
+runPauseable initial stepHandler eventHandler drawHandler isPaused = do
+    Just window <- currentWindow
+    Just doc <- currentDocument
+    Just canvas <- getElementById doc ("screen" :: JSString)
+
+    (sendEvent, getState) <- run initial stepHandler eventHandler drawHandler
+
+    onEvents canvas $ \event -> do
+        debugActive <- isDebugModeActive
+        when debugActive $ do
+            nextStatePaused <- isPaused <$> eventHandler event <$> getState
+            when (not nextStatePaused) $ setDebugModeActive False
+        sendEvent event
+
+    let picIfUnpaused state = case isPaused state of
+            True  -> drawHandler state
+            False -> pictures []
+        checkPaused = do
+            paused <- isPaused <$> getState
+            when (not paused) $ do
+                setDebugModeActive False
+                sweetAlert "error" "Sorry!" "Please pause before inspecting."
```

Comment:
If you auto-pause interactions, why not do the same with simulations and animations?  That would be much more convenient for the user.